src/python/pants/notes/2.19.x.md

```diff
@@ -1,7 +1,162 @@
+
+- [`cache_from` and `cache_to`](https://www.pantsbuild.org/v2.19/docs/docker#external-cache-storage-backends) for using external cache storage backends
+- [`output`](https://www.pantsbuild.org/v2.19/docs/reference-docker_image#codeoutputcode) to control how/where `buildx` outputs an image when running `pants package ::`
+
+#### JVM
+
+JVM lock-files are more reliable, handling [some cases](https://github.com/pantsbuild/pants/issues/20162) of Coursier not including dependencies when expected.
+
+Using [`shading_rules`](https://www.pantsbuild.org/v2.19/docs/reference-deploy_jar#codeshading_rulescode) is now supported on `deploy_jar` targets defined in directories of the build root.
+
+##### Kotlin
+
+The Kotlin analyzer now runs using the Zulu JDK, to be able to run natively on Apple Silicon (arm64 macOS).
+
+##### Java
+
+Dependency inference for Java now supports code using recently added syntax like `sealed class` & `permits`, `switch` expressions and `yield`.
+
+##### Scala
+
+Scala 3 is now supported when using [the BSP in an IDE](https://www.pantsbuild.org/v2.19/docs/jvm-overview#working-in-an-ide).
+
+#### OpenAPI
+
+[The new `pants.backend.experimental.openapi.lint.openapi_format` backend](https://www.pantsbuild.org/v2.19/docs/reference-openapi-format) supports running [openapi-format](https://github.com/thim81/openapi-format) to format `openapi_...` targets.
+
+#### Protobuf
+
+[The `pants.backend.codegen.protobuf.python` backend](https://www.pantsbuild.org/v2.19/docs/protobuf-python) now supports:
+
+- `protobuf_sources` targets that are part of multiple Python resolves, like `python_resolve=parametrize("first-resolve", "second-resolve")`.
+- `[python-infer].ambiguity_resolution = "by_source_root"` to infer dependencies without warnings or errors in more cases.
+
+Version 24.4 of [the `protoc` compiler](https://www.pantsbuild.org/v2.19/docs/reference-protoc) is included as a known version (but is not the default). It can be selected by setting `[protoc].version = "24.4"`. This version has native support for Apple Silicon (arm64 macOS).
+
+[Buf (`pants.backend.codegen.protobuf.lint.buf`)](https://www.pantsbuild.org/v2.19/docs/reference-buf) now correctly ignores `buf.work.yaml`, `buf.lock`, `buf.gen.lock` when running as a linter.
+
+#### Python
+
+**Deprecation**: Using the `platforms` field on `pex_binary` is now deprecated. It is not a recommended way to build for platforms other than the local one, as it is a lossy abbreviation of the target platform and often leads to binaries that don't work. To replace this, either [use `complete_platforms` to continue cross-compiling](https://www.pantsbuild.org/v2.19/docs/pex#setting-the-target-platforms-for-a-pex) or use [an environment matching the target platform](https://www.pantsbuild.org/v2.19/docs/environments).
+
+Pants now uses version 2.1.148 of the PEX CLI by default, which includes support for Python 3.12.
```

Comment:
I would add more clear instructions on how to setup 3.12. I spent some time trying to configure 3.12 along with 3.11, I had to explicitly set pip_version="23.2" because 23.1.2 only works with python<3.12

Reply:
Ah, thanks. That's a good idea. Before I update the text, I'd like to understand exactly what I should write.

It _seems_ to me that the default `pip_version = "23.1.2"` does work with Python 3.12? But I might be testing incorrectly:

```shell
python3.12 -m venv /tmp/venv312
. /tmp/venv312/bin/activate

python -m pip install -U pip==23.1.2

pip --version
#> pip 23.1.2 from /private/tmp/v312/lib/python3.12/site-packages/pip (python 3.12)

pip install pandas
# works, installs cp312 wheels
```

As such, I'm inclined to write something like:

```suggestion
Pants now uses version 2.1.148 of the PEX CLI by default, which includes support for Python 3.12. To use Pants with Python 3.12, you may need to adjust or remove any overrides you have for the default value of [`[pex-cli].version`](https://www.pantsbuild.org/v2.19/docs/reference-pex-cli#version) (and `known_versions`) and [`[python].pip_version`](https://www.pantsbuild.org/v2.19/docs/reference-python#pip_version).
```

What do you think?

Reply:
If you add python tests, it will fail with this error:
```
The Pip requested was pip==23.0.1 but it does not work with the interpreter selected which is CPython 3.12.1 at /home/grihabor/.pyenv/versions/3.12.1/bin/python3.12. Pip 23.0.1 requires Python <3.12,>=3.7.
```
Builtin resolve for pytest uses pip<23.2, so it doesn't work out of the box, you have to manually create pytest resolve and generate the lockfile with pip>=23.2

Reply:
I'm not sure it this is the right place to write the details about how to setup a resolve, but it might be worth adding a not about `mypy`, `flake8` and `pytest` - I can see they use pinned pip 3.1.2

Reply:
or maybe we should just generate new lockfiles for 2.19.x branch?

Reply:
> pip==23.0.1 

Ah, that's an even older version. Regenerating all of them to use a new lockfiles sounds good!

I've filed https://github.com/pantsbuild/pants/issues/20354 for using a newer pip, and investigating some improvements on the pex side too: https://github.com/pantsbuild/pex/pull/2314.

Reply:
Ah, okay, so the conclusion from those is I think Pants working perfectly with 3.12, we'll need to:

1. use a newer `pip_version` by default
2. update all the lockfiles to use this too

I think this is too much this late in the cycle, so I'll rewrite this:

```suggestion
Pants now uses version 2.1.148 of the PEX CLI by default, which includes support for Python 3.12. To use Python 3.12, you will need to set [`[python].pip_version`](https://www.pantsbuild.org/v2.19/docs/reference-python#pip_version) to `latest` or `23.2`, and ensure all of your tools use resolves generated with this new version (Pants' built in ones are not).
```

Reply:
lgtm, thanks!